contracts/TokenSugar.vy

```diff
@@ -89,27 +89,51 @@ def tokens(_limit: uint256, _offset: uint256, _account: address, \
     tokens: address[2] = [staticcall pool.token0(), staticcall pool.token1()]
 
     for i: uint256 in range(2):
-      if tokens[i] in seen:
+      token_index: uint256 = len(seen)
+      seen_count = len(seen)
+
+      # find index of seen token
+      for j: uint256 in range(0, MAX_TOKENS):
+        if j >= seen_count:
+          break
+        if seen[j] == tokens[i]:
+          token_index = j
+          break
+      
+      # if the token is already marked as emerging, skip
+      if token_index < seen_count and emerging[token_index]:
         continue
 
-      emerging: bool = False
-
       if self.v2_launcher.address != empty(address):
         launcher: IPoolLauncher = self.v2_launcher
         # check if pool is CL pool
         if pool_data[3] != empty(address):
           launcher = self.cl_launcher
 
-        # if pool is emerging and other token is pairable, set token as emerging
-        if staticcall launcher.emerging(pool_data[1]) > 0 and staticcall launcher.isPairableToken(tokens[1 - i]):
-          emerging = True
+        # if pool is emerging and token is not whitelisted, set token as emerging
+        if staticcall launcher.emerging(pool_data[1]) > 0 and not staticcall lp_shared.voter.isWhitelistedToken(tokens[i]):
```

Comment:
Was looking closer and another reason to consider a refactoring is that `_token()` already does a call to `voter.isWhitelistedToken()`

So I think if we pre-fetch the emerging token addresses, it will save that extra call as well.